crates/common/src/transaction.rs

```diff
@@ -0,0 +1,627 @@
+use stark_hash::{Felt, HashChain};
+
+use crate::{felt_bytes, prelude::*};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Transaction {
+    pub hash: TransactionHash,
+    pub variant: TransactionVariant,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum TransactionVariant {
+    DeclareV0(DeclareTransactionV0V1),
+    DeclareV1(DeclareTransactionV0V1),
+    DeclareV2(DeclareTransactionV2),
+    // Regenesis: deploy is a legacy variant and can be removed after regenesis.
+    Deploy(DeployTransaction),
+    DeployAccount(DeployAccountTransaction),
+    InvokeV0(InvokeTransactionV0),
+    InvokeV1(InvokeTransactionV1),
+    L1Handler(L1HandlerTransaction),
+}
```

Comment:
Should these instead be structs instead of newtype tuples? Would be a lot less extra types floating around..

```suggestion
pub enum TransactionVariant {
    DeclareV0 {
        pub class_hash: ClassHash,
        pub max_fee: Fee,
        pub nonce: TransactionNonce,
        pub sender_address: ContractAddress,
        pub signature: Vec<TransactionSignatureElem>,
    },
    ...
}
```

Reply:
Good idea.